### ISSUE TITLE
Field Slip Print Facility

### ISSUE THREAD

**Issue by mo-nathan** [MushroomObserver/mushroom-observer#1949]
Just use a generic template and the Field Slip QR code.  Can get more configurable later.  Based on playing with designed and community discussion have settled on 6 slips/page.  My understanding is that printers can collate in any desired way, so makes the most sense to do them sequentially on a page.  Should track the max number that has been printed, but allow value to be overriden.

**Comment by mo-nathan** [comment]
Using prawn is really easy!  I have a working version of a command line ruby script that now generates printable field slips.  See an example here: https://mushroomobserver.org/images/1661389.

**Comment by mo-nathan** [comment]
Gathered some stats on the current field slip generator.  My local system was able to generate a PDF with 6,000 field slips in 3 minutes 24 seconds or roughly 30 slip/second.  This seems reasonable, but does justify using a background job system.  If you assume at a large event that each participant will get a pack of 50 field slips you get to 10,000 slips with 200 participants.  I'm wondering what the upper limit should be.  In the current format the limit is 99,999 which on my local system should take less than an hour to generate.  However, it isn't clear how much of impact generating these files in the background will have on the production server.

**Comment by mo-nathan** [comment]
Decided not to worry about optimistic locking as well as providing explicit start numbers.  I think we should get this out and add these things only if needed.

**Comment by mo-nathan** [comment]
Nimmo enabled support for JavaScript updating of a Project's Field Slip job queue.  At this point I think the PR is ready to get merged.

**Comment by mo-nathan** [comment]
Draft of initial Field Slip documentation:

# MO Projects and Field Slips

## TL;DR (Too Long; Didn't Read):
You can now print your very own MO field slips for any MO project!  For example,

https://mushroomobserver.org/images/1661400

Note the QR code.  Scan it with your phone.  It will take you to the associated MO observation.

To get your own:

1. Create or edit an existing MO project.
2. Add a "Field Slip Prefix" to the project.  E.g., "NFAL" is the field slip prefix for the North Falmouth Fungi project.
3. From the Project page click, "Create Field Slip PDF".
4. Click "Create".
5. Download and print the PDF once it's "Done".
6. Cut apart the slips.
7. Find some mushrooms.
8. Take some notes on the slip.
9. Take some photos of the mushroom ideally including the slips.
10. Scan the QR code with your phone.
11. Click 'Create New Observation" on the resulting "Record Field Slip" page.
12. Fill out the MO create observation form and save it.
13. Keep the field slip with your collection.
14. Anyone can get back to that MO Observation any time by rescanning that QR code.
15. Rejoice in not having to write down the observation number or your login details on the field slip!

## Not Long Enough; Want to Read More

For those familiar with MO, Projects have been around for a long time.  However, over the last year or so we've been actively working to improve them with a focus on supporting forays that include vouchering such as the annual NAMA (North American Mycological Association) foray as well as regional or club forays such as the NEMF (NorthEastern Mycological Federation) foray.

To that end we have implemented a new feature we call "MO Field Slips".  MO Field Slips are intended to more conveniently associate collections and DNA samples with MO observations.  Field slips have of course been used for decades (centuries?) at forays with a more scientific bent to record details associated with collections that are intended for submission into fungaria or samples intended for DNA sequencing.  However, traditional field slips are a bit awkward to use with more modern online systems like MO and iNat because they require users to accurately and clearly provide enough information on the field slip to uniquely identify the relevant online observation.

MO Field Slips include a QR (Quick-Response) code that can be used to initiate the process of creating an observation on MO.  As a side effect that observation becomes forever associated with that QR code.  If such an observation already exists, the code will take you immediately to the associated observation.  These days most mobile phones can read QR codes automatically through their camera.


### Projects

To get started with your own MO Field Slips you need to be an admin or a member of an MO Project.  This can be an already existing project or you create a new one by going to the "Projects" link in the left-hand nav bar (https://mushroomobserver.org/projects) and then clicking "Add Project" in the upper right of that page (https://mushroomobserver.org/projects/new).  If you are already a member or admin of an MO project, you can go to that project page. Start by selecting "Your Projects" from the dropdown under your name in the upper right of any MO page.

To create your own MO Field Slips, the project must have a "Field Slip Prefix".  If you created the project or have been added as a project admin, you can set the Field Slip Prefix by clicking "Edit Project".  MO Projects can now be configured to let anyone add themselves to the project as a member.  An example of such an "open" project is the "Field Slip α-Test" project (https://mushroomobserver.org/projects/341) which has the field slip prefix "ATEST".

Field slip prefixes can only have English letters and Arabic numerals.  The letters will automatically be capitalized and must be unique across all existing MO projects to avoid confusion.  We recommend using the initials or an abbreviation of your local mushroom club or the project.  For example, the "North Falmouth Fungi" project uses the prefix "NFAL".

As a convenience, if you (or one of your friends) are the first person to scan a QR code for an open project, you (or they) are automatically added as a member.  Note that you must have a Mushroom Observer account for this to work.  You will be asked to login or create an account if needed.  Once you've done that, you may need to rescan the QR code to complete the process.  However, once a QR code is associated with an observation, you do not need to be logged in or have an account to reach that observation.

### Creating New MO Field Slips

When a project has a Field Slip Prefix, the prefix will be shown on the project page along with a button that says "Create Field Slip PDF".  If you click this button you will be able to initiate the creation of a PDF file with a set of as yet unused field slips for your project.  By default it will create a PDF with a single page containing 6 field slips.  This should take under a second to generate and once it's done you can down the PDF by clicking the file name on the page.

If you want to create a bunch of field slips, you can increase the number of pages.  Project members can create at most 60 field slips (10 pages) at one time.  Admins can create up to 12,000 field slips (2,000 pages!) at one time.  Such high numbers should only be used if you are planning on taking the results to a print shop to pre-print for a bunch of people, for example before a big foray event.

Once you have downloaded the PDF, you should print it out.  Using water proof or acid-free paper may be desirable, but is not required.  You will then need to cut up the pages.  Printing the PDF and cutting up the resulting pages is something print shops should be able to do for you and they can also bind them as separate notepads.  Once you have the separate field slips, they are ready to use.

### Using MO Field Slips

MO Field Slips can be used in the field, once you're back at the id room, or after the foray is done.  The first time a field slip is scanned it takes you to the "Record Field Slip" page on Mushroom Observer.  If it takes you to an observation instead then that field slip has already been scanned at least once.
The most common workflow is to click the "Create New Observation" button.  This will initiate the regular MO observation creation process.  If you complete that process, then the QR code for this field slip will in most cases be permanently associated with that observation.  You can test this by rescanning the QR code which should bring up the newly created observation.

If you do not complete the regular MO observation creation process, you can rescan the QR code and it will take you to a page just for that field slip.  You can then "Edit this field slip".  Editing a field slip allows you to associate the field slip with any observation in the system or create a new observation to associate with that QR code.  The field slip edit interface, gives you the opportunity to associate the field slip with the last observation you have viewed in MO.  If the observation it shows you is not the one you want to associate with the QR code, find the observation you want to associate with it (make sure you click on that observation to view just that observation), and return to the Edit Field Slip page either by rescanning the QR code or by reloading the page.  Once the observation you intend to use is showing, click "Use Last Observation".  Note that this option is also available when you first scan the QR code.

If you end up associating a field slip with the wrong observation, you can change it scanning the QR code and then clicking the appropriate "Field Slip" which should be shown on the right side of the observation page.  This will take you to a page specifically for that field slip with the option to "Edit this field slip".

Finally, if you want to start over completely with a field slip, you can go to the field slip page as described above, and click "Destroy this field slip".  This will cause MO to forget about this field slip and you can start over by scanning the QR code on the field slip.

### What If I Don't Have a Mobile Phone or the QR code?

There are situations, such as folks doing DNA sequencing, where you may have the code (e.g., ATEST-00001), but not the QR code.  Folks without mobile phones can also read the code straight off the field slip.  The easiest way to use the field slip code to get to the observation is by going to the URL `https://mushroomobserver.org/qr/<code>` (e.g., https://mushroomobserver.org/qr/ATEST-00001).  If the code has not yet been associated with an observation in MO, then it will let you go ahead and record the new MO Field Slip as described above.  We are working to provide a search option for field code as well so you don't have to mess with creating a custom URL.

**Comment by nimmolo** [comment]
Now that's a writeup
